reindexer/reindex_worker/src/main/scala/uk/ac/wellcome/platform/reindex_worker/services/ReindexService.scala

```diff
@@ -28,47 +30,53 @@ class ReindexService @Inject()(dynamoDbClient: AmazonDynamoDB,
     dynamoConfig = dynamoConfig
   )
 
-  def runReindex(reindexJob: ReindexJob): Future[List[Unit]] =
-    Future {
-      info(s"ReindexService running $reindexJob")
-
-      val table = Table[ReindexRecord](dynamoConfig.table)
-
-      val index = table.index(indexName = dynamoConfig.index)
+  def runReindex(reindexJob: ReindexJob): Future[List[Unit]] = {
+    info(s"ReindexService running $reindexJob")
+    val table = Table[ReindexRecord](dynamoConfig.table)
+    for {
+      index: SecondaryIndex[ReindexRecord] <- Future.fromTry(Try {
+        table.index(indexName = dynamoConfig.index)
+      })
 
-      // We start by querying DynamoDB for every record in the reindex shard
-      // that has an out-of-date reindexVersion.  If a shard was especially
-      // large, this might cause out-of-memory errors -- in practice, we're
-      // hoping that the shards/individual records are small enough for this
-      // not to be a problem.
-      val results: List[Either[DynamoReadError, ReindexRecord]] =
+      results: List[Either[DynamoReadError, ReindexRecord]] <- Future {
         Scanamo.exec(dynamoDbClient)(
           index.query(
             'reindexShard -> reindexJob.shardId and
               KeyIs('reindexVersion, LT, reindexJob.desiredVersion)
           )
         )
-
-      val outdatedRecords: List[ReindexRecord] = results.map {
-        case Left(err: DynamoReadError) => {
-          warn(s"Failed to read Dynamo records: $err")
-          throw GracefulFailureException(
-            new RuntimeException(s"Error in the DynamoDB query: $err")
-          )
-        }
-        case Right(r: ReindexRecord) => r
       }
 
+      // We start by querying DynamoDB for every record in the reindex shard
+      // that has an out-of-date reindexVersion.  If a shard was especially
+      // large, this might cause out-of-memory errors -- in practice, we're
+      // hoping that the shards/individual records are small enough for this
+      // not to be a problem.
+      outdatedRecords: List[ReindexRecord] = results.map { extractRecord(_) }
+
       // Then we PUT all the records.  It might be more efficient to do a
       // bulk update, but this will do for now.
-      outdatedRecords
-        .map { record: ReindexRecord =>
-          val updatedRecord =
-            record.copy(reindexVersion = reindexJob.desiredVersion)
-          versionedDao.updateRecord[ReindexRecord](updatedRecord)
-        }
-        .map { _ =>
-          ()
-        }
+      _ <- outdatedRecords.map {
```

Comment:
Don't you need a Future.sequence here? This creates a List[Future]], so I don't think this waits on the completion of individual updates. Are there any tests that check what happens if an individual update fails? If not we should add one

Reply:
Yep, yep. Test added, and I’ve got the Future.sequence stuff working as well.